collect/stressRelief.go

```diff
@@ -0,0 +1,307 @@
+package collect
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/dgryski/go-wyhash"
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/logger"
+	"github.com/honeycombio/refinery/metrics"
+)
+
+type StressReliever interface {
+	Start() error
+	UpdateFromConfig(cfg config.StressReliefConfig) error
+	Recalc()
+	StressLevel() uint
+	Stressed() bool
+	GetSampleRate(traceID string) (rate uint, keep bool, reason string)
+}
+
+type MockStressReliever struct{}
+
+func (m *MockStressReliever) Start() error                                         { return nil }
+func (m *MockStressReliever) UpdateFromConfig(cfg config.StressReliefConfig) error { return nil }
+func (m *MockStressReliever) Recalc()                                              {}
+func (m *MockStressReliever) StressLevel() uint                                    { return 0 }
+func (m *MockStressReliever) Stressed() bool                                       { return false }
+func (m *MockStressReliever) GetSampleRate(traceID string) (rate uint, keep bool, reason string) {
+	return 1, false, ""
+}
+
+// hashSeed is a random value to seed the hash generator for the sampler.
+// We want it to be a constant that's the same across all nodes so that they
+// all make the same sampling decisions during stress relief.
+const hashSeed = 34527861234
+
+type StressReliefMode int
+
+const (
+	Never StressReliefMode = iota
+	Monitor
+	Always
+)
+
+type StressRelief struct {
+	mode            StressReliefMode
+	activateLevel   uint
+	deactivateLevel uint
+	sampleRate      uint64
+	upperBound      uint64
+	stressLevel     uint
+	reason          string
+	stressed        bool
+	belowMin        bool
+	minDuration     time.Duration
+	RefineryMetrics metrics.Metrics `inject:"metrics"`
+	Logger          logger.Logger   `inject:""`
+	Done            chan struct{}
+
+	algorithms map[string]func(string, string) float64
+	calcs      []StressReliefCalculation
+	lock       sync.RWMutex
+}
+
+func (s *StressRelief) Start() error {
+	s.Logger.Debug().Logf("Starting StressRelief system")
+	defer func() { s.Logger.Debug().Logf("Finished starting StressRelief system") }()
+
+	s.algorithms = map[string]func(string, string) float64{
+		"linear":  s.linear,
+		"sqrt":    s.sqrt,
+		"square":  s.square,
+		"sigmoid": s.sigmoid,
+	}
+
+	s.calcs = []StressReliefCalculation{
+		{Numerator: "collector_peer_queue_length", Denominator: "PEER_CAP", Algorithm: "sqrt", Reason: "CacheCapacity (peer)"},
+		{Numerator: "collector_incoming_queue_length", Denominator: "INCOMING_CAP", Algorithm: "sqrt", Reason: "CacheCapacity (incoming)"},
+		{Numerator: "libhoney_peer_queue_length", Denominator: "PEER_BUFFER_SIZE", Algorithm: "sqrt", Reason: "PeerBufferSize"},
+		{Numerator: "libhoney_upstream_queue_length", Denominator: "UPSTREAM_BUFFER_SIZE", Algorithm: "sqrt", Reason: "UpstreamBufferSize"},
+		{Numerator: "memory_heap_allocation", Denominator: "MEMORY_MAX_ALLOC", Algorithm: "sigmoid", Reason: "MaxAlloc"},
```

Comment:
It looks like only sqrt and sigmoid are used, do we need definitions for linear and square?

Reply:
Again, this is intended for future configuration and tuning. I will add some more comments to it.